src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -182,9 +183,11 @@ inline bool ShenandoahHeap::check_cancelled_gc_and_yield(bool sts_active) {
   }
 }
 
-inline void ShenandoahHeap::clear_cancelled_gc() {
+inline void ShenandoahHeap::clear_cancelled_gc(bool clear_oom_handler) {
   _cancelled_gc.set(CANCELLABLE);
-  _oom_evac_handler.clear();
+  if (clear_oom_handler) {
+    _oom_evac_handler.clear();
```

Comment:
Prior to these changes, the gc cancellation flag was always set because of an allocation failure, so the flag was cleared on a safepoint (final update refs). Now we have a case where we've "cancelled" a gc, but not because we've run out of memory so we don't need to clear the oom evac handler.

Reply:
Ok. But I think the OOM handler still needs to be cleared at the safepoint.

Reply:
Ah yes, sorry. The default value for the boolean is true (to clear the oom handler). I've updated the other callers to pass the argument explicitly.